### PR TITLE
[AutoTVM] Added an autotuning Config Library to store autotune results

### PR DESCRIPTION
**Pull request by mbaret**
This implements the first stage of an offline AutoTVM cache, somewhat like an offline TopHub. Auto-tuning can be performed in 'jobs' which are a series of sequentially tuned tasks. Data describing the platform on which the job was performed + when it took place can also be saved into the library.
    
This is initial infrastructure to allow for more advanced management of the auto-tuning process. It includes a mechanism to skip tasks that are already tuned in the library.
    
For further information, see the tutorial: 'tutorials/autotvm/tune_with_config_library.py'.

RFC is here: https://github.com/dmlc/tvm/issues/4150

### CODE REVIEW

**Comment by merrymercy** [comment]
Sorry, I am busy this week. I will try to review later this week. @mbarrett97 You can invite other people to review this in the meanwhile.

**Comment by mbarrett97** [comment]
@comaniac @eqy Could you please take a look at this?

File: python/tvm/autotvm/config_library.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+            no_indent_id = int(json_obj[no_indent_index_start+2:no_indent_index_stop])
+            no_indent_obj = PyObj_FromPtr(no_indent_id)
+            no_indent_json = json.dumps(no_indent_obj.value)
+            json_obj = json_obj.replace(
+                '"{}"'.format(self.marker.format(no_indent_id)),
+                no_indent_json,
+            )
+            offset = no_indent_index_stop + 2
+
+        return json_obj
```
Comment (comaniac) [inline python/tvm/autotvm/config_library.py]: If the purpose of implementing this class is just to put multiple JSON objects to the same file, I would suggest using one nested JSON object. For example:

Current:
out.json
{ 1st config }
{ 2nd config }

Proposed:
out.json
[{1st config},
 {2nd config}]

File: python/tvm/autotvm/config_library.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+from pathlib import Path
+
+import numpy as np
+
+from . import record
+from .task import ApplyHistoryBest
+
+
+class ConfigLibraryException(Exception):
+    pass
```
Comment (comaniac) [inline python/tvm/autotvm/config_library.py]: Why we need this?

File: python/tvm/autotvm/tuner/tuner.py [hunk 3]
```diff
@@ -86,8 +87,7 @@ def update(self, inputs, results):
             result for measurement
         """
 
-
-    def tune(self, n_trial, measure_option, early_stopping=None, callbacks=()):
+    def tune(self, n_trial, measure_option, early_stopping=None, callbacks=(), allow_skipping=True):
```
Comment (comaniac) [inline python/tvm/autotvm/tuner/tuner.py]: I think the default value of `allow_skipping` should be `False`, which means users have to intentionally skip tasks.

File: tests/python/unittest/test_autotvm_config_library.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+from tvm.autotvm.record import encode
+
+from test_autotvm_common import get_sample_records
+
+
+def get_random_configs(n):
+    records = get_sample_records(n)
+    random_configs = []
+    r = random.Random()
+    r.seed(42)
```
Comment (comaniac) [inline tests/python/unittest/test_autotvm_config_library.py]: You can hard code random configs, but do not use random generator in the unit tests (fixing the seed number doesn't guarantee the result is deterministic on different machines) sicne it is hard to catch the problem by all means.

File: tests/python/unittest/test_autotvm_tuning_job.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+from test_autotvm_common import get_sample_records
+
+
+class MockTuner:
+    pass
+
+
+def test_log_configs():
+    r = random.Random()
+    r.seed(42)
```
Comment (comaniac) [inline tests/python/unittest/test_autotvm_tuning_job.py]: ditto

File: python/tvm/autotvm/config_library.py (hunk 1 above)
Comment (mbarrett97) [reply python/tvm/autotvm/config_library.py]: I store the configs currently as 
{ 
  [workload_str]: config,
  [workload_str]: config,
}
to make it simpler to query by workload. The issue I was finding was when using the JSON indent option, the configs got massively expanded as they are themselves heavily nested objects. This encoder makes it so the configs ignore the indent value and are printed on one line. I'm very open to an alternative method here as it seems quite messy!

File: tests/python/unittest/test_autotvm_config_library.py (hunk 4 above)
Comment (mbarrett97) [reply tests/python/unittest/test_autotvm_config_library.py]: Understood, I'll change this!

File: python/tvm/autotvm/config_library.py (hunk 2 above)
Comment (mbarrett97) [reply python/tvm/autotvm/config_library.py]: Think I just forgot to get rid of this, will be removed.

File: python/tvm/autotvm/config_library.py (hunk 1 above)
Comment (comaniac) [reply python/tvm/autotvm/config_library.py]: Why we need to indent the JSON file if we didn't expect people to read it?

Comment (mbarrett97) [reply python/tvm/autotvm/config_library.py]: If no one cares about human readability I'm happy to drop this. I was mostly trying to preserve the appearance of the existing format where it's 1 config per line.

**Comment by mbarrett97** [comment]
@comaniac Regarding your concerns from the RFC, specifically the possibility of some 'task pass' infrastructure, I think they're outside of the scope of the PR. I can try and accommodate system in a later PR once it is fully developed.

**Comment by mbarrett97** [comment]
@comaniac @merrymercy @eqy Can I get further comment/an approving review on this?

**Comment by merrymercy** [comment]
@mbarrett97 any update?

**Pull request closed without merging**